legate/core/runtime.py

```diff
@@ -135,172 +148,206 @@ def update_free_fields(self) -> None:
-                runtime.max_field_reuse_frequency + ratio - 1
-            ) // ratio
-        else:
-            self.match_frequency = runtime.max_field_reuse_frequency
 
     def destroy(self) -> None:
         self.free_fields = deque()
-        self.freed_fields = []
+
+    def try_reuse_field(self) -> Optional[tuple[Region, int]]:
+        if len(self.free_fields) > 0:
+            return self.free_fields.popleft()
+        else:
+            return None
 
     def allocate_field(self) -> tuple[Region, int]:
-        # Increment our match counter
-        self.match_counter += 1
-        # If the match counter equals our match frequency then do an exchange
-        if self.match_counter == self.match_frequency:
-            # This is where the rubber meets the road between control
-            # replication and garbage collection. We need to see if there
-            # are any freed fields that are shared across all the shards.
-            # We have to test this deterministically no matter what even
-            # if we don't have any fields to offer ourselves because this
-            # is a collective with other shards. If we have any we can use
-            # the first one and put the remainder on our free fields list
-            # so that we can reuse them later. If there aren't any then
-            # all the shards will go allocate a new field.
-            local_freed_fields = self.freed_fields
-            # The match now owns our freed fields so make a new list
-            # Have to do this before dispatching the match
-            self.freed_fields = list()
-            match = FieldMatch(self, local_freed_fields)
-            # Dispatch the match
-            self.runtime.dispatch(match)
-            # Put it on the deque of outstanding matches
-            self.matches.append(match)
-            # Reset the match counter back to 0
-            self.match_counter = 0
+        result = self.try_reuse_field()
+        if result is not None:
+            region_manager = self.runtime.find_region_manager(result[0])
+            region_manager.increase_field_count()
+            return result
+        region_manager = self.runtime.find_or_create_region_manager(self.shape)
+        return region_manager.allocate_field(self.field_size)
+
+    def free_field(
+        self, region: Region, field_id: int, ordered: bool = False
+    ) -> None:
+        self.free_fields.append((region, field_id))
+        region_manager = self.runtime.find_region_manager(region)
+        destroy = region_manager.decrease_field_count()
+        if destroy:
+            self.runtime.destroy_region_manager(
+                self.shape, region, unordered=not ordered
+            )
+
+    def remove_all_fields(self, region: Region) -> None:
+        free_fields: Deque[tuple[Region, int]] = deque()
+        for pair in self.free_fields:
+            if pair[0] != region:
+                free_fields.append(pair)
+        self.free_fields = free_fields
```

Comment:
can simplify and avoid the explicit type (I think) by using a generator:
```suggestion
        self.free_fields = deque(f for f in self.free_fields if f[0] != region)
```

Reply:
https://github.com/nv-legate/legate.core/pull/364/commits/6060fe1900dbace84a5f3eabcf3508f678a71e46